{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a36b685b",
   "metadata": {},
   "source": [
    "# Exploring the set of human k-mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cd39542",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import time\n",
    "import json\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from numpy.typing import NDArray\n",
    "from Bio import SeqIO\n",
    "import tqdm.notebook as tqdm\n",
    "\n",
    "from humprot.helpers import sym2int\n",
    "from humprot.helpers import get_kmers_from_sequences\n",
    "from humprot.helpers import count_kmers_in_seqs, build_i2s_s2i_maps\n",
    "from humprot.trees import MultiKmerTree\n",
    "from humprot.constants import AA_LIST_STD20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ce075da",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATDIR = \"../data\"\n",
    "OUTDIR = \"../out/human_kmer_exploration\"\n",
    "IMGDIR = os.path.join(OUTDIR, \"images\")\n",
    "\n",
    "os.makedirs(OUTDIR, exist_ok=True)\n",
    "os.makedirs(IMGDIR, exist_ok=True)\n",
    "\n",
    "human_mktree_fpath = os.path.join(DATDIR, \"human_multikmer_trees\", \"human_noXU_mktree_25.npz\")\n",
    "human_fasta_fpath = os.path.join(DATDIR, \"human_proteins_noXU.faa\")\n",
    "\n",
    "KMAX = 25\n",
    "KS_ARR = np.arange(1, KMAX + 1)\n",
    "\n",
    "if not os.path.isfile(human_mktree_fpath):\n",
    "    raise FileNotFoundError(human_mktree_fpath)\n",
    "\n",
    "if not os.path.isfile(human_fasta_fpath):\n",
    "    raise FileNotFoundError(human_fasta_fpath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d0c70f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "AA_LIST = AA_LIST_STD20\n",
    "MASK = len(AA_LIST)\n",
    "NUM_AAS = len(AA_LIST)\n",
    "\n",
    "int2sym_map, sym2int_map = build_i2s_s2i_maps(AA_LIST, mask=MASK)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907c634e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = MultiKmerTree.load(human_mktree_fpath)\n",
    "print(f\"Loaded tree from file: {human_mktree_fpath}\")\n",
    "tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02b112f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "layer_sizes = tree.get_layer_sizes()\n",
    "layer_totals = tree.get_layer_value_sums()\n",
    "\n",
    "print(\"layer sizes:\", layer_sizes)\n",
    "print(\"layer totals:\", layer_totals)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e034aac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot number of unique realized k-mers as a function of k\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "ax.semilogy(\n",
    "    KS_ARR, layer_sizes, \".-\",\n",
    ")\n",
    "\n",
    "ax.set_xlabel(f\"$k$\")\n",
    "ax.set_ylabel(\"count\")\n",
    "ax.set_title(\"Number of unique realized human $k$-mers\")\n",
    "\n",
    "ax.grid(which='major', color='gray', linestyle='-', alpha=0.6)\n",
    "ax.set_xlim(0, KMAX + 1)\n",
    "\n",
    "plt.savefig(f\"{IMGDIR}/k_vs_num_unique_kmers.png\", bbox_inches=None)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb0ab2cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot coverage of kmer space as a function of k\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "densities = np.log(layer_sizes) - KS_ARR * np.log(NUM_AAS)\n",
    "ax.plot(\n",
    "    KS_ARR, densities, \".-\"\n",
    ")\n",
    "\n",
    "ax.set_xlabel(f\"$k$\")\n",
    "ax.set_ylabel(\"coverage rate\")\n",
    "ax.set_title(\"Coverage of all possible $k$-mers\")\n",
    "\n",
    "ax.grid(which='major', color='gray', linestyle='-', alpha=0.6)\n",
    "ax.grid(which='minor', color='gray', linestyle='--', alpha=0.3)\n",
    "ax.set_xlim(0, KMAX + 1)\n",
    "\n",
    "plt.savefig(f\"{IMGDIR}/k_vs_coverage_percent.png\", bbox_inches=None)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ebc2f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot total number of kmers, with repeats, as a function of k\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "nuniq_kmers = layer_totals\n",
    "\n",
    "ax.plot(\n",
    "    KS_ARR, nuniq_kmers, \".-\"\n",
    ")\n",
    "\n",
    "ax.set_xlabel(f\"$k$\")\n",
    "ax.set_ylabel(\"count\")\n",
    "ax.set_title(\"Total number of human $k$-mers (with repeats)\")\n",
    "\n",
    "ax.grid(which='major', color='gray', linestyle='-', alpha=0.6)\n",
    "ax.grid(which='minor', color='gray', linestyle='--', alpha=0.3)\n",
    "ax.set_xlim(0, ax.get_xlim()[1])\n",
    "\n",
    "plt.savefig(f\"{IMGDIR}/k_vs_total_num_kmers.png\", bbox_inches=None)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f98fdac",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuniq_kmers = layer_totals\n",
    "\n",
    "# Iterate over the layers of the tree\n",
    "for i, node_idxs in enumerate(tree.iternodes_in_layers()):\n",
    "    k = i + 1\n",
    "    density = densities[i]\n",
    "    print(f\"k={k}\")\n",
    "    #~~~ Plot frequencies of realized k-mers\n",
    "    counts = tree._get_value_at_node(node_idxs) \n",
    "\n",
    "    fig, ax = plt.subplots(1, 1)\n",
    "    ax.hist(\n",
    "        counts, \n",
    "        bins=30, log=True,\n",
    "        label=f\"total={format(len(node_idxs), \",\")}\"\n",
    "    )\n",
    "    ax.set_xlabel(f\"num occurrences in proteome\")\n",
    "    ax.set_ylabel(f\"count\")\n",
    "    ax.set_title(\"Human {}-mer frequencies\\n{} unique (log density {:.3g})\".format(\n",
    "        k, format(len(node_idxs), \",\"), density\n",
    "    ))\n",
    "    ax.legend()\n",
    "\n",
    "    plt.savefig(f\"{IMGDIR}/frequencies_{k}mers.png\")\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab06fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuniq_kmers = layer_totals\n",
    "\n",
    "# Iterate over the layers of the tree\n",
    "for i, node_idxs in enumerate(tree.iternodes_in_layers()):\n",
    "    k = i + 1\n",
    "    density = densities[i]\n",
    "    print(f\"k={k}\")\n",
    "    \n",
    "    #~~~ Plot number of children\n",
    "    num_children = tree._get_numchildren_at_node(node_idxs)\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1)\n",
    "    ax.hist(\n",
    "        num_children, \n",
    "        bins=30, log=True,\n",
    "        label=f\"total={format(len(node_idxs), \",\")}\"\n",
    "    )\n",
    "    ax.set_xlabel(f\"num continuations\")\n",
    "    ax.set_ylabel(f\"count\")\n",
    "    ax.set_title(\"Human {}-mer possible continuations $k\\\\to k+1$\".format(k))\n",
    "\n",
    "    ax.legend()\n",
    "\n",
    "    plt.savefig(f\"{IMGDIR}/frequencies_{k}mers.png\")\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf98b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuniq_kmers = layer_totals\n",
    "\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "ax.set_xlim(0, KMAX + 1)\n",
    "# Iterate over the layers of the tree\n",
    "to_plot = [2, 4, 8, 16, 24]\n",
    "means = np.zeros(KS_ARR.shape)\n",
    "mins = np.zeros(KS_ARR.shape)\n",
    "maxs = np.zeros(KS_ARR.shape)\n",
    "continuations = np.zeros([1 + NUM_AAS, KMAX], dtype=int)\n",
    "for i, node_idxs in enumerate(tree.iternodes_in_layers()):\n",
    "    k = i + 1\n",
    "    density = densities[i]\n",
    "    print(f\"k={k}\")\n",
    "    \n",
    "    #~~~ Plot number of children\n",
    "    num_children = tree._get_numchildren_at_node(node_idxs)\n",
    "    print(num_children.min(), num_children.max(), num_children.mean())\n",
    "\n",
    "    mins[i] = num_children.min()\n",
    "    maxs[i] = num_children.max()\n",
    "    means[i] = num_children.mean()\n",
    "\n",
    "    from collections import Counter\n",
    "    counts = Counter(num_children)\n",
    "    counts = np.array([counts.get(key, 0) for key in range(len(continuations))])\n",
    "    continuations[:,i] = counts\n",
    "    \n",
    "    if k not in to_plot:\n",
    "        continue\n",
    "    ax.hist(\n",
    "        num_children, \n",
    "        bins=10, log=True,\n",
    "        label=f\"{k}\", alpha=0.2, align=\"mid\",\n",
    "    )\n",
    "ax.set_xlabel(f\"num continuations\")\n",
    "ax.set_ylabel(f\"count\")\n",
    "ax.set_title(\"Human k-mer possible continuations $k\\\\to k+1$\")\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left', title=\"$k$\")\n",
    "\n",
    "plt.savefig(f\"{IMGDIR}/frequencies_kmers.png\")\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ccc2371",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "ax.plot(KS_ARR, mins, 'r.')\n",
    "ax.plot(KS_ARR, maxs, 'r.')\n",
    "ax.plot(KS_ARR, means, 'k-.')\n",
    "ax.set_xlabel(f\"$k$\")\n",
    "ax.set_ylabel(f\"mean number of possible continuations\")\n",
    "ax.set_title(f\"Average continuation ability\")\n",
    "\n",
    "\n",
    "plt.savefig(f\"{IMGDIR}/frequencies_kmers.png\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df03cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "from matplotlib.colors import LogNorm\n",
    "\n",
    "sc = ax.imshow(\n",
    "    continuations[:,:-1], \n",
    "    cmap='viridis',\n",
    "    norm=LogNorm(),        # Log normalization\n",
    "    aspect='equal',        # square cells like matshow\n",
    "    origin='lower',        # same as matshow default\n",
    "    interpolation='none',  # no smoothing\n",
    ")\n",
    "\n",
    "cbar = fig.colorbar(sc)\n",
    "cbar.ax.set_ylabel(\"$k$-mer count\")\n",
    "\n",
    "ax.set_xlabel(f\"$k$\")\n",
    "ax.set_ylabel(f\"unique continuations available\")\n",
    "ax.set_title(\"Human k-mer possible continuations $k\\\\to k+1$\")\n",
    "\n",
    "xts = np.arange(0, KMAX, 2)\n",
    "yts = np.arange(0, NUM_AAS + 1, 2)\n",
    "ax.set_xticks(\n",
    "    xts, labels=[f\"{i + 1}\" for i in xts]\n",
    ")\n",
    "ax.set_yticks(\n",
    "    yts, labels=[f\"{i}\" for i in yts]\n",
    ")\n",
    "\n",
    "\n",
    "plt.savefig(f\"{IMGDIR}/continuation_possibilities.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41316204",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5640d597",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in tree.iternodes_in_layers(max_depth=25):\n",
    "    print(len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09d968cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_func = lambda nidx: tree._get_numchildren_at_node(nidx)\n",
    "for depth, x in enumerate(tree.iternodes_in_layers(max_depth=25, node_func=node_func)):\n",
    "    print(f\"Depth {depth}:\", np.mean(x), np.median(x), np.min(x), np.max(x))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "287aeec1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b7dca0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the sequence data\n",
    "human_id2seq = {}\n",
    "for record in SeqIO.parse(human_fasta_fpath, \"fasta\"):\n",
    "    seqid = record.id\n",
    "    seq = sym2int(record.seq, sym2int_map)\n",
    "    human_id2seq[seqid] = seq\n",
    "\n",
    "print(f\"Loaded {len(human_id2seq)} human protein sequences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e05b25",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
